### PR TITLE
Gateways can register with the federation

### PR DESCRIPTION
**Pull request by justinmoon**
* Replacement for https://github.com/fedimint/minimint/pull/200
* Removes `client.json`
* Saves gateway config in federation and client databases

### CODE REVIEW

File: minimint/src/net/api.rs [hunk 1]
```diff
@@ -126,6 +126,18 @@ fn server_endpoints() -> &'static [ApiEndpoint<MinimintConsensus<rand::rngs::OsR
                 Ok(tx_status)
             }
         },
+        api_endpoint! {
+            "/list_gateways",
+            async |minimint: &MinimintConsensus<rand::rngs::OsRng>, _v: serde_json::Value| -> Vec<LightningGateway> {
```
Comment (Maan2003) [inline minimint/src/net/api.rs:131]:
```suggestion
            async |minimint: &MinimintConsensus<rand::rngs::OsRng>, _: ()| -> Vec<LightningGateway> {
```

Comment (justinmoon) [reply minimint/src/net/api.rs:131]: Suggested change doesn't compile w/ error:

```
async |minimint: &MinimintConsensus<rand::rngs::OsRng>, _: ()| -> Vec<LightningGateway> {
                                                        ^ no rules expected this token in macro call
```

Comment (Maan2003) [reply minimint/src/net/api.rs:131]: ah `_` is not supported by the macro
you can just try `_v`

Comment (justinmoon) [reply minimint/src/net/api.rs]: Updated with `_v: ()`

**Comment by justinmoon** [comment]
Unclear why integration tests are failing: https://github.com/fedimint/minimint/runs/7365495170?check_suite_focus=true#step:6:4939

I think I'm going to revise the last 2 commits and just have `lightningd` always start without the plugin running. Then call `lightning-cli plugin start` (or rpc equivalent) wherever we need it. We can fork `cln-rpc` to add RPC struct for this (and stop) so that rust tests can completely direct the plugin on per-test basis.

File: minimint/src/consensus/mod.rs [hunk 2]
```diff
@@ -513,6 +514,19 @@ where
...
+            .find_by_prefix(&LightningGatewayKeyPrefix)
+            .map(|res| res.expect("DB error").1)
+            .collect()
+    }
+
+    pub fn register_gateway(&self, gateway: LightningGateway) {
+        self.db
+            .insert_entry(&LightningGatewayKey(gateway.node_pub_key), &gateway)
+            .expect("DB error");
+    }
```
Comment (elsirion) [inline minimint/src/consensus/mod.rs:529]: What's the reasoning behind putting these endpoints into the federation consensus itself instead of the LN module?

File: docs/database.md [hunk 3]
```diff
@@ -21,8 +21,10 @@ The Database is split into different key spaces based on prefixing that can be u
 
 | Name                  | Prefix | Key                              | Value                           |
 |-----------------------|--------|----------------------------------|---------------------------------|
-| Pending Transactions  | `0x01`   | Transaction ID (sha256, 32bytes) | Transaction                     |
-| Accepted Transactions | `0x02`   | Transaction ID (sha256, 32bytes) | Confirmation epoch, Transaction |
+| Pending Transactions  | `0x01` | Transaction ID (sha256, 32bytes) | Transaction                     |
+| Accepted Transactions | `0x02` | Transaction ID (sha256, 32bytes) | Confirmation epoch, Transaction |
+| Drop Peer             | `0x03` | Peer ID (u16)                    |                                 |
+| Lightning Gateways    | `0x04` | Node Pubkey (PublicKey)          | LightningGateway                |
```
Comment (elsirion) [inline docs/database.md:27]: If we were to move the API endpoints to the LN module the prefix also had to change.

File: client/client-lib/src/lib.rs [hunk 4]
```diff
@@ -505,6 +505,21 @@ impl Client<UserClientConfig> {
...
+    /// Notify gateway that we've escrowed tokens they can claim by routing our payment and wait
+    /// for them to do so
+    pub async fn await_outgoing_contract_execution(&self, contract_id: ContractId) -> Result<()> {
+        let gateway = self.fetch_gateway().await?;
+
+        let future = reqwest::Client::new()
+            .post(&format!("{}/pay_invoice", gateway.api))
+            .json(&contract_id)
+            .send();
+        minimint_api::task::timeout(Duration::from_secs(15), future)
```
Comment (elsirion) [inline client/client-lib/src/lib.rs:530]: The timeout should be coordinated with the gateway imo. Many LN payments will take longer, the question is how long the GW will try. We should only time out if the gateway said it would try for time t and >t time passed.

File: ln-gateway/src/bin/ln_gateway.rs [hunk 5]
```diff
@@ -51,9 +41,20 @@ async fn generate_config(workdir: &Path, ln_client: &mut ClnRpc) {
     };
     let node_pub_key = secp256k1::PublicKey::from_slice(&node_pub_key_bytes.to_vec()).unwrap();
 
+    // Write gateway config
+    let gateway_cfg = GatewayClientConfig {
+        client_config: federation_client_cfg.clone(),
+        redeem_key: kp_fed,
+        timelock_delta: 10,
+        node_pub_key,
+        api: String::from("http://127.0.0.1:8080"), // FIXME: don't hard-code this
```
Comment (elsirion) [inline ln-gateway/src/bin/ln_gateway.rs:50]: I guess that will be another PR?

File: ln-gateway/src/bin/ln_gateway.rs [hunk 6]
```diff
@@ -51,9 +41,20 @@ async fn generate_config(workdir: &Path, ln_client: &mut ClnRpc) {
...
+    let gw_cfg_file_path: PathBuf = workdir.join("gateway.json");
+    let gw_cfg_file = std::fs::File::create(gw_cfg_file_path).expect("Could not create cfg file");
+    serde_json::to_writer_pretty(gw_cfg_file, &gateway_cfg).unwrap();
+
     // Write user config
     let client_cfg = UserClientConfig(federation_client_cfg);
-
     let client_cfg_file_path: PathBuf = workdir.join("client.json");
     let client_cfg_file =
         std::fs::File::create(client_cfg_file_path).expect("Could not create cfg file");
```
Comment (elsirion) [inline ln-gateway/src/bin/ln_gateway.rs]: Is this still necessary with the gateway registration code?

**Review by elsirion — COMMENTED** [review]
I left some questions and nits. CI doesn't seem to run right now, maybe you need to rebase on #275?

File: ln-gateway/src/bin/ln_gateway.rs (hunk 6 above)
Comment (elsirion) [reply ln-gateway/src/bin/ln_gateway.rs]: I think we can just `cp` the `federation_client.json` to `client.json` right? Or do we need to make the newtype transparent for serde for that to work? (in which case I'd propose to do that)

File: client/client-lib/src/lib.rs (hunk 4 above)
Comment (justinmoon) [reply client/client-lib/src/lib.rs:530]: Made an [issue](https://github.com/fedimint/minimint/issues/301) for this. Will address in a separate PR.

File: ln-gateway/src/bin/ln_gateway.rs (hunk 5 above)
Comment (justinmoon) [reply ln-gateway/src/bin/ln_gateway.rs:49]: Yes. This is fine for now.

File: ln-gateway/src/bin/ln_gateway.rs (hunk 6 above)
Comment (justinmoon) [reply ln-gateway/src/bin/ln_gateway.rs]: I got rid of `federation_client.json`. Now we just have `client.json` which can be parsed as `UserClientConfig` or `ClientConfig`.

File: client/client-lib/src/lib.rs [hunk 7]
```diff
@@ -83,17 +84,26 @@ pub struct PaymentParameters {
 }
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub struct UserClientConfig {
-    pub client_config: ClientConfig,
-    pub gateway: LightningGateway,
-}
+pub struct UserClientConfig(pub ClientConfig);
```
Comment (NicolaLS) [inline client/client-lib/src/lib.rs:87]: Now that `UserClientConfig` just wraps `ClientConfig` do we still have a reason for this new-type ?
Or could we just do `type UserClient = Client<ClientConfig>` instead of `..Client<UserClientConfig>`

Comment (justinmoon) [reply client/client-lib/src/lib.rs:87]: Yea we could do that now. Will give it a try.

Comment (elsirion) [reply client/client-lib/src/lib.rs:87]: Iirc we discussed that and came to the conclusion that the newtypes let us expose certain functionalities for e.g. only the user client, which might avoid certain kinds of user errors (trying to pay via a GW with a GW, potentially leading to a cyclic dependency).

Comment (justinmoon) [reply client/client-lib/src/lib.rs:87]: Ahh that's right

File: minimint/src/consensus/mod.rs (hunk 2 above)
Comment (justinmoon) [reply minimint/src/consensus/mod.rs]: That was a mistake. Fixed now.

File: docs/database.md (hunk 3 above)
Comment (justinmoon) [reply docs/database.md]: Done.

File: scripts/lib.sh [hunk 8]
```diff
@@ -37,6 +37,11 @@ function await_server_on_port() {
 # Function for killing processes stored in FM_PID_FILE
 function kill_minimint_processes {
   kill $(cat $FM_PID_FILE | sed '1!G;h;$!d') #sed reverses the order here
-  pkill "ln_gateway";
+  pkill "ln_gateway" || true;
```
Comment (justinmoon) [inline scripts/lib.sh:40]: This was breaking CI because the gateway plugin is no longer run by the rust tests. So we just kill the gateway if it's running and don't fail if it isn't running.

**Review by elsirion — APPROVED** [review]
Awesome! Finally we don't need to worry about getting GW configs to the user anymore.
